src/hooks/useTrackListSearchParams.ts

```diff
@@ -51,51 +52,62 @@ function isValidSearchParameter<T>(value: unknown, options: T[]): value is T {
   return options.includes(value as T);
 }
 
-function getValidatedSearchParam<T>(
+function getNumberParam(
   searchParams: URLSearchParams,
   key: string,
-  validator: (value: unknown) => value is T,
-): T | null;
-
-function getValidatedSearchParam<T>(
-  searchParams: URLSearchParams,
-  key: string,
-  validator: (value: unknown) => value is T,
-  defaultValue: T,
-): T;
+  defaultValue: number,
+) {
+  return pipe(
+    O.fromNullable(searchParams.get(key)),
+    O.flatMap((value) => {
+      const n = Number(value);
+      return isNaN(n) ? O.None : O.Some(n);
+    }),
+    O.getWithDefault(defaultValue),
+  );
+}
 
 function getValidatedSearchParam<T>(
   searchParams: URLSearchParams,
   key: string,
-  validator: (value: unknown) => value is T,
-  defaultValue: T | null = null,
-): T | null {
-  const param = searchParams.get(key);
-  if (!param) return defaultValue;
-  if (!validator(param)) return defaultValue;
-
-  return param;
+  validOptions: T[],
+): O.Option<T> {
+  return pipe(
+    O.fromNullable(searchParams.get(key)),
+    O.flatMap((value) =>
+      isValidSearchParameter(value, validOptions) ? O.Some(value) : O.None,
+    ),
+  );
 }
 
 type SortAndOrderShape =
```

Comment:
Consider moving the SortAndOrderShape type to the top of the file or even extracting it into a separate types.ts module if it’s used across multiple places.
This would improve readability and maintainability, especially as the file grows.